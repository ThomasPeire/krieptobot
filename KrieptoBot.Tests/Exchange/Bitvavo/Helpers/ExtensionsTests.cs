using FluentAssertions;
using KrieptoBot.Infrastructure.Bitvavo.Dtos;
using KrieptoBot.Infrastructure.Bitvavo.Extensions;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NUnit.Framework;
using Snapshooter.NUnit;
using System;
using System.Collections.Generic;
using System.Linq;

namespace KrieptoBot.Tests.Exchange.Bitvavo.Helpers
{
    public class ExtensionsTests
    {
        private IEnumerable<AssetDto> _assets;
        private IEnumerable<BalanceDto> _balances;
        private IEnumerable<CandleDto> _candles;
        private IEnumerable<MarketDto> _markets;
        private IEnumerable<OrderDto> _orders;
        private IEnumerable<TradeDto> _trades;

        [SetUp]
        public void Setup()
        {
            InitAssets();
            InitBalances();
            InitMarkets();
            InitCandles();
            InitOrders();
            InitTrades();
        }

        private void InitAssets()
        {
            var assetsJson = System.IO.File.ReadAllText(@"./Mocks/Bitvavo/Data/assets.json");
            _assets = JsonConvert.DeserializeObject<IEnumerable<AssetDto>>(assetsJson);
        }

        private void InitBalances()
        {
            var balancesJson = System.IO.File.ReadAllText(@"./Mocks/Bitvavo/Data/balances.json");
            _balances = JsonConvert.DeserializeObject<IEnumerable<BalanceDto>>(balancesJson);
        }

        private void InitMarkets()
        { 
            var marketsJson = System.IO.File.ReadAllText(@"./Mocks/Bitvavo/Data/markets.json");
            _markets = JsonConvert.DeserializeObject<IEnumerable<MarketDto>>(marketsJson);
        }

        private void InitCandles()
        {
            var candlesJson = System.IO.File.ReadAllText(@"./Mocks/Bitvavo/Data/candles_btc-eur.json");
            var deserializedCandles = JsonConvert.DeserializeObject(candlesJson) as JArray;
            _candles = deserializedCandles.Select(x =>
                new CandleDto
                {
                    TimeStamp = DateTime.UnixEpoch.AddMilliseconds(x.Value<long>(0)),
                    Open = x.Value<decimal>(1),
                    High = x.Value<decimal>(2),
                    Low = x.Value<decimal>(3),
                    Close = x.Value<decimal>(4),
                    Volume = x.Value<decimal>(5),
                });
        }

        private void InitOrders()
        {
            var ordersJson = System.IO.File.ReadAllText(@"./Mocks/Bitvavo/Data/orders_btc-eur.json");
            _orders = JsonConvert.DeserializeObject<IEnumerable<OrderDto>>(ordersJson);
        }

        private void InitTrades()
        {
            var tradesJson = System.IO.File.ReadAllText(@"./Mocks/Bitvavo/Data/trades_btc-eur.json");
            _trades = JsonConvert.DeserializeObject<IEnumerable<TradeDto>>(tradesJson);
        }

        [Test]
        public void ConvertToKrieptoBotModel_ShouldConvert_BitvavoAssetToKrieptoBotAsset()
        {
            var result = _assets.First().ConvertToKrieptoBotModel();

            result.Should().MatchSnapshot();
        }

        [Test]
        public void ConvertToKrieptoBotModel_ShouldConvert_BitvavoBalanceToKrieptoBotBalance()
        {
            var result = _balances.First().ConvertToKrieptoBotModel();

            result.Should().MatchSnapshot();
        }

        [Test]
        public void ConvertToKrieptoBotModel_ShouldConvert_BitvavoCandleToKrieptoBotCandle()
        {
            var result = _candles.First().ConvertToKrieptoBotModel();

            result.Should().MatchSnapshot();
        }

        [Test]
        public void ConvertToKrieptoBotModel_ShouldConvert_BitvavoMarketToKrieptoBotMarket()
        {
            var result = _markets.First().ConvertToKrieptoBotModel();

            result.Should().MatchSnapshot();
        }

        [Test]
        public void ConvertToKrieptoBotModel_ShouldConvert_BitvavoOrderToKrieptoBotOrder()
        {
            var result = _orders.First().ConvertToKrieptoBotModel();

            result.Should().MatchSnapshot();
        }

        [Test]
        public void ConvertToKrieptoBotModel_ShouldConvert_BitvavoTradeToKrieptoBotTrade()
        {
            var result = _trades.First().ConvertToKrieptoBotModel();

            result.Should().MatchSnapshot();
        }

    }
}