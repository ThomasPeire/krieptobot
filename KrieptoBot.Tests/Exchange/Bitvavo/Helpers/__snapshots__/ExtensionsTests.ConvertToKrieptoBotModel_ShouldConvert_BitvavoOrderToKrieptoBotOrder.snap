{
  "OrderId": "4a7bd126-2d21-4918-96dc-0c8f51760a0b",
  "Market": "BTC-EUR",
  "Created": "2021-03-11T08:05:31.466Z",
  "Updated": "2021-03-11T08:05:31.466Z",
  "Status": "filled",
  "Side": "sell",
  "OrderType": "limit",
  "Amount": 0.00241469,
  "AmountRemaining": 0.0,
  "Price": 45928.0,
  "AmountQuote": 0.0,
  "AmountQuoteRemaining": 0.0,
  "OnHold": 0.0,
  "OnHoldCurrency": "BTC",
  "TriggerPrice": 0.0,
  "TriggerAmount": 0.0,
  "TriggerType": null,
  "TriggerReference": null,
  "FilledAmount": 0.00241469,
  "FilledAmountQuote": 110.90188232,
  "FeePaid": 0.17188232,
  "FeeCurrency": "EUR",
  "Fills": [
    {
      "Id": "c9d0abf0-9377-4b1a-ab06-a840f730dc43",
      "Timestamp": "2021-03-11T08:23:55.833Z",
      "Amount": 0.00241469,
      "Price": 45928.0,
      "Taker": false,
      "Fee": 0.17188232,
      "FeeCurrency": "EUR",
      "Settled": true
    }
  ],
  "SelfTradePrevention": "decrementAndCancel",
  "Visible": false,
  "TimeInForce": "GTC",
  "PostOnly": false,
  "DisableMarketProtection": false
}
